odoo/tools/float_utils.py

```diff
@@ -71,47 +86,31 @@ def float_round(value, precision_digits=None, precision_rounding=None, rounding_
     # the order of magnitude of the value, to tip the tie-break in the right
     # direction.
     # Credit: discussion with OpenERP community members on bug 882036
-
-    normalized_value = value / rounding_factor # normalize
-    sign = math.copysign(1.0, normalized_value)
     epsilon_magnitude = math.log(abs(normalized_value), 2)
-    epsilon = 2**(epsilon_magnitude-52)
-
-    # TIE-BREAKING: UP/DOWN (for ceiling[resp. flooring] operations)
-    # When rounding the value up[resp. down], we instead subtract[resp. add] the epsilon value
-    # as the approximation of the real value may be slightly *above* the
-    # tie limit, this would result in incorrectly rounding up[resp. down] to the next number
-    # The math.ceil[resp. math.floor] operation is applied on the absolute value in order to
-    # round "away from zero" and not "towards infinity", then the sign is
-    # restored.
-
-    if rounding_method == 'UP':
-        normalized_value -= sign*epsilon
-        rounded_value = math.ceil(abs(normalized_value)) * sign
-
-    elif rounding_method == 'DOWN':
-        normalized_value += sign*epsilon
-        rounded_value = math.floor(abs(normalized_value)) * sign
-
-    # TIE-BREAKING: HALF-EVEN
-    # We want to apply HALF-EVEN tie-breaking rules, i.e. 0.5 rounds towards closest even number.
-    elif rounding_method == 'HALF-EVEN':
-        rounded_value = math.copysign(builtins.round(normalized_value), normalized_value)
-
-    # TIE-BREAKING: HALF-DOWN
-    # We want to apply HALF-DOWN tie-breaking rules, i.e. 0.5 rounds towards 0.
-    elif rounding_method == 'HALF-DOWN':
-        normalized_value -= math.copysign(epsilon, normalized_value)
-        rounded_value = round(normalized_value)
-
-    # TIE-BREAKING: HALF-UP (for normal rounding)
-    # We want to apply HALF-UP tie-breaking rules, i.e. 0.5 rounds away from 0.
-    else:
-        normalized_value += math.copysign(epsilon, normalized_value)
-        rounded_value = round(normalized_value)     # round to integer
-
-    result = rounded_value * rounding_factor # de-normalize
-    return result
+    # `epsilon_magnitude - 52` would be the minimal size, but we increase it to make it more
+    # tolerant of inaccuracies accumulated after multiple floating point operations
+    epsilon = 2**(epsilon_magnitude - 50)
```

Comment:
I don't know if it's genius or madness :smiling_face_with_tear:

Reply:
I mean, it's either this or require devs to call `float_round` after every single addition. subtraction, multiplication or division, or "pre-round" every float using a slightly higher precision before doing the eventual round.

Reply:
I know 🙈 Even, it’s still removing 2 from the python mantissa in an arbitrary way.  Btw, don’t forget to launch the l10n build because I’m pretty sure it will not be green

Reply:
Not sure why it's not listed, but it seems to be built & green: https://runbot.odoo.com/runbot/batch/1432802/build/61694739